crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -2515,3 +2517,51 @@ fn range_size_hint(range: &impl RangeBounds<TxNumber>) -> Option<usize> {
     };
     end.checked_sub(start).map(|x| x as _)
 }
+
+/// Limits on how long pruning can run before it's forced to stop and thereby yield the
+/// [`DatabaseProviderRW`] hook.
+#[derive(Debug, Default, Clone, Copy, Constructor)]
+pub struct PruneLimit {
+    /// Maximum entries to delete from the database in one prune job.
+    segment_limit: Option<usize>,
+    /// The max time one prune job can run.
+    job_timeout: Option<Duration>,
+}
+
+impl PruneLimit {
+    /// Returns the maximum entries that can be deleted from the database in one prune job. `None`
+    /// is equivalent to unlimited segments.
```

Comment:
I think there's confusion between entries and segments:
- Entries are basically rows in the database, and we don't want to delete more than X of them in one pruner run
- Segments are parts of the database that represent one or more tables, and we don't want to limit them as long as the number of deleted entries is respected